From 780888d3e64dd19c6441c1eb76ed98a286885ec0 Mon Sep 17 00:00:00 2001
From: Kuninori Morimoto <kuninori.morimoto.gx@renesas.com>
Date: Fri, 5 Jul 2013 15:45:26 +0900
Subject: [PATCH 617/675] ARM: shmobile: bockw: add USB Function support

Bock-W USB1 (CN29) can be USB Host/Func by SW98/SW99 settings.
USB Func will be enabled if CONFIG_USB_RENESAS_USBHS_UDC[_MODULE]
was selected

Signed-off-by: Kuninori Morimoto <kuninori.morimoto.gx@renesas.com>
---
 arch/arm/mach-shmobile/board-bockw.c | 48 +++++++++++++++++++++++++++++++++---
 1 file changed, 45 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-shmobile/board-bockw.c b/arch/arm/mach-shmobile/board-bockw.c
index 68748e0..5a9797b 100644
--- a/arch/arm/mach-shmobile/board-bockw.c
+++ b/arch/arm/mach-shmobile/board-bockw.c
@@ -91,6 +91,16 @@ static void __iomem *fpga;
  * # amixer set "LINEOUT Mixer DACL" on
  */
 
+/*
+ * USB
+ *
+ * USB1 (CN29) can be Host/Function
+ *
+ *		Host	Func
+ * SW98		0	2
+ * SW99		0	3
+ */
+
 /* Dummy supplies, where voltage doesn't matter */
 static struct regulator_consumer_supply dummy_supplies[] = {
 	REGULATOR_SUPPLY("vddvario", "smsc911x"),
@@ -406,7 +416,39 @@ static struct platform_device ak4554_ssi78_device = {
 	.id	= 1,
 };
 
-static struct rcar_phy_platform_data usb_phy_platform_data __initdata;
+/* USB */
+#if	defined(CONFIG_USB_RENESAS_USBHS_UDC) || \
+	defined(CONFIG_USB_RENESAS_USBHS_UDC_MODULE)
+
+	/* USB1 is Func */
+	static struct renesas_usbhs_platform_info usbhs_info = {
+		.platform_callback = {
+			/* .get_id	will be set in register function */
+			/* .power_ctrl	will be set in register function */
+		},
+		.driver_param = {
+			.buswait_bwait	= 4,
+		},
+	};
+
+	#define USB_PHY_SETTING {.port1_func = 1, .ovc_pin[1].active_high = 1,}
+	#define PIN_MAP_MUX_GROUP_USB1 \
+		PIN_MAP_MUX_GROUP_DEFAULT("renesas_usbhs", "pfc-r8a7778", \
+					  "usb1_ovc", "usb1")
+	#define R8A7778_ADD_USB_FUNC_DEVICE_IF_POSSIBLE \
+		r8a7778_add_usb_func_device(&usbhs_info)
+#else
+	/* USB1 is Host */
+	#define USB_PHY_SETTING { }
+	#define PIN_MAP_MUX_GROUP_USB1 \
+		PIN_MAP_MUX_GROUP_DEFAULT("ehci-platform", "pfc-r8a7778", \
+					  "usb1", "usb1")
+	#define R8A7778_ADD_USB_FUNC_DEVICE_IF_POSSIBLE
+#endif
+
+static struct rcar_phy_platform_data usb_phy_platform_data __initdata =
+	USB_PHY_SETTING;
+
 
 static struct rcar_vin_platform_data vin_platform_data __initdata = {
 	.flags	= RCAR_VIN_BT656,
@@ -466,8 +508,7 @@ static const struct pinctrl_map bockw_pinctrl_map[] = {
 	/* USB */
 	PIN_MAP_MUX_GROUP_DEFAULT("ehci-platform", "pfc-r8a7778",
 				  "usb0", "usb0"),
-	PIN_MAP_MUX_GROUP_DEFAULT("ehci-platform", "pfc-r8a7778",
-				  "usb1", "usb1"),
+	PIN_MAP_MUX_GROUP_USB1,
 	/* VIN0 */
 	PIN_MAP_MUX_GROUP_DEFAULT("r8a7778-vin.0", "pfc-r8a7778",
 				  "vin0_clk", "vin0"),
@@ -601,6 +642,7 @@ static void __init bockw_init(void)
 static void __init bockw_init_late(void)
 {
 	r8a7778_add_usb_host_device();
+	R8A7778_ADD_USB_FUNC_DEVICE_IF_POSSIBLE;
 }
 
 static const char *bockw_boards_compat_dt[] __initdata = {
-- 
1.8.2

